lib/bakeAmbientOcclusion.js

```diff
@@ -0,0 +1,616 @@
+    var bufferDataByAccessor = {};
+
+    for (var accessorID in accessors) {
+        if (accessors.hasOwnProperty(accessorID)) {
+            var accessor = accessors[accessorID];
+            bufferDataByAccessor[accessorID] = readAccessor(gltf, accessor);
+        }
+    }
+
+    var raytracerScene = {
+        bufferDataByAccessor: bufferDataByAccessor,
+        numberSamples: defaultValue(options.numberSamples, 16),
+        rayDepth: defaultValue(options.rayDepth, 1.0), // TODO: compute dynamic default ray depth?
+        triangleSoup: [],
+        aoBufferByPrimitive: {},
+        nearCull: 0.5 / options.resolution
+    };
+
+    // TODO: currently assuming each primitive appears in the scene once. this is not true. figure out what to do.
+
+    // Generate all the world transform matrices
+    NodeHelpers.computeFlatTransformScene(scene, gltf.nodes);
+
+    var parameters = {
+        meshes: gltf.meshes,
+        resolution: options.resolution,
+        raytracerScene: raytracerScene
+    };
+
+    // Generate triangle soup over the whole scene
+    var rootNodeNames = scene.nodes;
+    var allNodes = gltf.nodes;
+    for (var nodeID in rootNodeNames) {
+        if (rootNodeNames.hasOwnProperty(nodeID)) {
+            var rootNodeName = rootNodeNames[nodeID];
+            NodeHelpers.depthFirstTraversal(allNodes[rootNodeName], allNodes, addNodeToSoup, parameters);
+        }
+    }
+
+    return raytracerScene;
+}
+
+function addNodeToSoup(parameters, node) {
+    var transform = node.extras._pipeline.flatTransform;
+    var mesheIDs = node.meshes;
+    for (var meshIndex in mesheIDs) {
+        if (mesheIDs.hasOwnProperty(meshIndex)) {
+            var meshID = mesheIDs[meshIndex];
+            var meshData = parameters.meshes[meshID];
+            var primitives = meshData.primitives;
+            for (var primitiveID in primitives) {
+                if (primitives.hasOwnProperty(primitiveID)) {
+                    processPrimitive(parameters, meshID + '_' + primitiveID, primitives[primitiveID], transform);
+                }
+            }
+        }
+    }
+}
+
+function processPrimitive(parameters, meshPrimitiveID, primitive, transform) {
+    var raytracerScene = parameters.raytracerScene;
+    var bufferDataByAccessor = raytracerScene.bufferDataByAccessor;
+    var indices = bufferDataByAccessor[primitive.indices].data;
+    var positions = bufferDataByAccessor[primitive.attributes.POSITION].data;
+    var numTriangles = indices.length / 3;
+    primitive.extras._pipeline.transform = transform;
+
+    var resolution = parameters.resolution;
+
+    var aoBuffer = {
+        resolution: resolution,
+        samples: new Array(resolution * resolution).fill(0.0),
+        count: new Array(resolution * resolution).fill(0.0)
+    };
+
+    raytracerScene.aoBufferByPrimitive[meshPrimitiveID] = aoBuffer;
+
+    // Read each triangle's Cartesian3s using the index buffer
+    for (var i = 0; i < numTriangles; i++) {
```

Comment:
Not all primitives are triangles, so you need to check the primitive's mode before you can make this assumption.